fontforgeexe/charinfo.c

```diff
@@ -3909,27 +3994,92 @@ static void CIFillup(CharInfo *ci) {
     CI_DoHideUnusedPair(ci);
     CI_DoHideUnusedSingle(ci);
 
-    bits = SFGetAlternate(sc->parent,sc->unicodeenc,sc,true);
+	GGadget *cotf = GWidgetGetControl(ci->gw,CID_ComponentTextField);
+	GGadget *codcb = GWidgetGetControl(ci->gw,CID_ComponentDefaultCB);
+	GGadget *coim = GWidgetGetControl(ci->gw,CID_ComponentInterpMsg);
+	GGadget *cola = GWidgetGetControl(ci->gw,CID_ComponentChangeMsg);
+
+    bits = bits2 = SFGetAlternate(sc->parent,sc->unicodeenc,NULL,true);
     GGadgetSetTitle8(GWidgetGetControl(ci->gw,CID_ComponentMsg),
 	bits==NULL ? _("No components") :
 	hascomposing(sc->parent,sc->unicodeenc,sc) ? _("Accented glyph composed of:") :
 	    _("Glyph composed of:"));
     if ( bits==NULL ) {
 	ubuf[0] = '\0';
 	GGadgetSetTitle(GWidgetGetControl(ci->gw,CID_Components),ubuf);
+    GGadgetSetTitle(cotf,ubuf);
     } else {
-	unichar_t *temp = malloc(11*u_strlen(bits)*sizeof(unichar_t));
+	unichar_t *temp = malloc(18*u_strlen(bits)*sizeof(unichar_t));
 	unichar_t *upt=temp;
 	while ( *bits!='\0' ) {
-	    sprintf(buffer, "U+%04x ", *bits );
+	    sprintf(buffer, "U+%04x (", *bits );
+	    uc_strcpy(upt,buffer);
+	    upt += u_strlen(upt);
+	    if (iscombining(*bits)) {
+	        *upt = 0x25CC; // DOTTED CIRCLE “◌”
+	        upt += 1;
+	    }
+	    *upt = *bits;
+	    upt += 1;
+	    sprintf(buffer, ") ");
 	    uc_strcpy(upt,buffer);
 	    upt += u_strlen(upt);
+
 	    ++bits;
 	}
 	upt[-1] = '\0';
 	GGadgetSetTitle(GWidgetGetControl(ci->gw,CID_Components),temp);
 	free(temp);
     }
+    if (bits2 == NULL) {
+        bits2 = (unichar_t*) L"\0";
```

Comment:
Don't do this, `wchar_t` has an implementation defined width (i.e. on windows it's 2 bytes, not 4 bytes). Given the use case below, something like this should work:

```c
if (d_ptr == NULL) d_ptr = bits2;

int d_length = d_ptr ? u_strlen(d_ptr) : 0;
```

Especially since you check for `d_ptr` being null in the while loop